src/librustc/ty/util.rs

```diff
@@ -1116,6 +1124,8 @@ fn needs_drop_raw<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
 
         ty::TyClosure(def_id, ref substs) => substs.upvar_tys(def_id, tcx).any(needs_drop),
 
+        ty::TyGenerator(..) => true,
```

Comment:
That's a pessimization (I didn't notice), but I suppose you don't want to force the interior? A comment would be welcome.